src/graphic/Fast3D/gfx_pc.cpp

```diff
@@ -944,6 +968,63 @@ static void import_texture(int i, int tile) {
     }
 }
 
+static void import_texture_mask(int i, int tile) {
+    uint32_t tmem_index = rdp.texture_tile[tile].tmem_index;
+    RawTexMetadata metadata = rdp.loaded_texture[tmem_index].raw_tex_metadata;
+
+    auto maskIter = masked_textures.find(metadata.name);
+    if (maskIter == masked_textures.end()) {
+        return;
+    }
+
+    const uint8_t* orig_addr = maskIter->second.mask;
+
+    if (orig_addr == nullptr) {
+        return;
+    }
+
+    TextureCacheKey key = { orig_addr, {}, 0, 0, 0 };
+
+    if (gfx_texture_cache_lookup(i, key)) {
+        return;
+    }
+
+    uint32_t width = rdp.texture_tile[tile].line_size_bytes;
+    uint32_t height =
+        rdp.loaded_texture[rdp.texture_tile[tile].tmem_index].orig_size_bytes / rdp.texture_tile[tile].line_size_bytes;
+    switch (rdp.texture_tile[tile].siz) {
+        case G_IM_SIZ_4b:
+            width *= 2;
+            break;
+        case G_IM_SIZ_8b:
+        default:
+            break;
+        case G_IM_SIZ_16b:
+            width /= 2;
+            break;
+        case G_IM_SIZ_32b:
+            width /= 4;
+            break;
+    }
+
+    for (uint32_t texIndex = 0; texIndex < width * height; texIndex++) {
+        uint8_t masked = orig_addr[texIndex];
+        if (masked) {
+            tex_upload_buffer[4 * texIndex + 0] = 0;
+            tex_upload_buffer[4 * texIndex + 1] = 0;
+            tex_upload_buffer[4 * texIndex + 2] = 0;
+            tex_upload_buffer[4 * texIndex + 3] = 0xFF;
+        } else {
+            tex_upload_buffer[4 * texIndex + 0] = 0;
+            tex_upload_buffer[4 * texIndex + 1] = 0;
+            tex_upload_buffer[4 * texIndex + 2] = 0;
+            tex_upload_buffer[4 * texIndex + 3] = 0;
+        }
+    }
```

Comment:
This is something I want to fix, but currently all textures are treated as RGBA32. It was easiest to just work with this for now.

Reply:
All cpu modified textures, or all textures period?

Reply:
All textures period. Every texture is converted to RGBA32 before being given to the APIs, and the APIs can only handle RGBA32. Even HD textures are being converted. I think this is done because not all textures map 1-1 between N64 and modern APIs, but it still could be optimized. Especially in this case, where the mask texture has literally 1 bit on information per texel but becomes 32 bits wide.

This would be an overhaul well outside the scope of the PR.

Reply:
Yeah that sounds good to me, especially given that it was existing behaviour. I do know that the easiest textures to upload to the backends are RGB. Definitely out of scope.